Env(name="hisat2", base="bioconda", packages=["hisat2", "samtools"])


rule hisat2_all:
    input: "{dir}.ht2/{:runs:}.sorted.bam.bai"
    output: touch("{dir}.ht2/all")


rule hisat2:
    """
    For hisat we always assume a pre-build index as providing SNPs and haplotypes
    etc is beyond this pipelines scope.

    TODO: allow choosing different references
    """
    message: "Hisat2: mapping "
    input:
        pairs = "{dir}/{source}.{: pairnames :}.fq.gz",
        index = "{:dir.references:}/hisat_GRCh38_snp_tran"
    output:
        bam   = temp("{dir}.ht2/{source}.bam"),
        stats = "{dir}.ht2/{source}.stats",
        unal  = "{dir}.ht2RMgrhc38/{source}.{:pairnames:}.fq.gz"
    log:
                "{dir}.ht2/{source}.log"
    params:
        mem   = icfg.mem("16G"),
        unal  = "{dir}.ht2RMgrhc38/{source}"
    threads:
        16
    conda:
        "hisat2"
    shell:
        "hisat2"
        " -1 {input.pairs[0]}"
        " -2 {input.pairs[1]}"
        " -x {input.index}/genome_snp_tran"
        " --new-summary"
        " --summary-file {output.stats}"
        " --downstream-transcriptome-assembly" #  require longer anchors for de novo splice
        " -p {threads} "
        " --un-conc-gz {params.unal}"
        " 2>{log}"
        " | samtools view -b -o {output.bam} -;"
        "mv {params.unal}.1 {output.unal[0]};"
        "mv {params.unal}.2 {output.unal[1]};"

